openmls/src/schedule/psk.rs

```diff
@@ -178,30 +179,88 @@ pub struct PreSharedKeyId {
 }
 
 impl PreSharedKeyId {
-    /// Create a new `PreSharedKeyID`
+    /// Construct a `PreSharedKeyID` with a random nonce.
     pub fn new(
         ciphersuite: Ciphersuite,
         rand: &impl OpenMlsRand,
         psk: Psk,
     ) -> Result<Self, CryptoError> {
-        Ok(Self {
+        let psk_nonce = rand
+            .random_vec(ciphersuite.hash_length())
+            .map_err(|_| CryptoError::InsufficientRandomness)?
+            .into();
+
+        Ok(Self { psk, psk_nonce })
+    }
+
+    /// Construct an external `PreSharedKeyID`.
+    pub fn external(psk_id: Vec<u8>, psk_nonce: Vec<u8>) -> Self {
+        let psk = Psk::External(ExternalPsk::new(psk_id));
+
+        Self {
             psk,
-            psk_nonce: rand
-                .random_vec(ciphersuite.hash_length())
-                .map_err(|_| CryptoError::InsufficientRandomness)?
-                .into(),
-        })
+            psk_nonce: psk_nonce.into(),
+        }
+    }
+
+    /// Construct a resumption `PreSharedKeyID`.
+    pub fn resumption(
+        usage: ResumptionPskUsage,
+        psk_group_id: GroupId,
+        psk_epoch: GroupEpoch,
+        psk_nonce: Vec<u8>,
+    ) -> Self {
+        let psk = Psk::Resumption(ResumptionPsk::new(usage, psk_group_id, psk_epoch));
+
+        Self {
+            psk,
+            psk_nonce: psk_nonce.into(),
+        }
     }
 
-    /// Return the PSK
+    /// Return the PSK.
     pub fn psk(&self) -> &Psk {
         &self.psk
     }
 
-    /// Return the PSK nonce
+    /// Return the PSK nonce.
     pub fn psk_nonce(&self) -> &[u8] {
         self.psk_nonce.as_slice()
     }
+
+    /// Save this `PreSharedKeyId` in the keystore.
+    ///
+    /// Note: The nonce is not saved as it must be present in the messages when processing PSKs.
+    pub fn write_to_key_store<KeyStore: OpenMlsKeyStore>(
+        &self,
+        backend: &impl OpenMlsCryptoProvider<KeyStoreProvider = KeyStore>,
+        ciphersuite: Ciphersuite,
+        psk: &[u8],
+    ) -> Result<(), KeyStore::Error> {
```

Comment:
This should probably retunr a `PskError`, which is either a key store error or a library error due to the serialization.